ldms/src/sampler/llnl/lustre_client.c

```diff
@@ -0,0 +1,641 @@
+                goto out1;
+        llite->name = strdup(llite_name);
+        if (llite->name == NULL)
+                goto out2;
+        snprintf(path_tmp, PATH_MAX, "%s/%s", basedir, llite_name);
+        llite->path = strdup(path_tmp);
+        if (llite->path == NULL)
+                goto out3;
+        snprintf(path_tmp, PATH_MAX, "%s/stats", llite->path);
+        llite->stats_path = strdup(path_tmp);
+        if (llite->stats_path == NULL)
+                goto out4;
+        llite->fs_name = strdup(llite_name);
+        if (llite->fs_name == NULL)
+                goto out5;
+        if (strtok_r(llite->fs_name, "-", &state) == NULL) {
+                log_fn(LDMSD_LWARNING, SAMP" unable to parse filesystem name from \"%s\"\n",
+                       llite->fs_name);
+                goto out6;
+        }
+        llite->general_metric_set = llite_general_create(producer_name, producer_name,
+                                                         llite->fs_name, llite->name,
+                                                         component_id);
+        if (llite->general_metric_set == NULL)
+                goto out6;
+        rbn_init(&llite->llite_tree_node, llite->name);
+
+        return llite;
+out6:
+        free(llite->fs_name);
+out5:
+        free(llite->stats_path);
+out4:
+        free(llite->path);
+out3:
+        free(llite->name);
+out2:
+        free(llite);
+out1:
+        return NULL;
+}
+
+static void llite_destroy(struct llite_data *llite)
+{
+        log_fn(LDMSD_LDEBUG, SAMP" llite_destroy() %s\n", llite->name);
+        llite_general_destroy(llite->general_metric_set);
+        free(llite->fs_name);
+        free(llite->stats_path);
+        free(llite->path);
+        free(llite->name);
+        free(llite);
+}
+
+static void llites_destroy()
+{
+        struct rbn *rbn;
+        struct llite_data *llite;
+
+        while (!rbt_empty(&llite_tree)) {
+                rbn = rbt_min(&llite_tree);
+                llite = container_of(rbn, struct llite_data,
+                                   llite_tree_node);
+                rbt_del(&llite_tree, rbn);
+                llite_destroy(llite);
+        }
+}
+
+/* List subdirectories in LLITE_PATH to get list of
+   LLITE names.  Create llite_data structures for any LLITES that we
+   have not seen, and delete any that we no longer see. */
+static void llites_refresh()
+{
+        struct dirent *dirent;
+        DIR *dir;
+        struct rbt new_llite_tree;
+
+        log_fn(LDMSD_LDEBUG, SAMP" rescanning llite directory\n");
+        rbt_init(&new_llite_tree, string_comparator);
+
+        /* Make sure we have llite_data objects in the new_llite_tree for
+           each currently existing directory.  We can find the objects
+           cached in the global llite_tree (in which case we move them
+           from llite_tree to new_llite_tree), or they can be newly allocated
+           here. */
+
+        dir = opendir(LLITE_PATH);
+        if (dir == NULL) {
```

Comment:
This warning should be issued exactly once before a successful sampling occurs; it needs a static guard variable. The guard can be reset to it's initial state (warning message allowed) after a successful sample. In the case where /proc disappears, we do not want the log flooded with redundant warnings.

Reply:
Done.